### ISSUE TITLE
Cinnamon Panel Too Wide / Icons Disappear/Overflow

### ISSUE THREAD

**Issue by basic6** [linuxmint/cinnamon#5042]
The panel gets wider and applets overlap when there's not enough space. Having a couple of windows open, the window list area will overlap the workspace list area, so it's not possible to switch to those windows that overlap workspace buttons (window button is in the foreground but when clicked, the workspace button behind it is activated). Furthermore, the whole panel grows instead of always staying at full width of the first screen. The first few workspace buttons disappear behind window buttons and the clock at the right is outside of the screen:
![workspace 1_010](https://cloud.githubusercontent.com/assets/2535481/12609325/5b2c9b2c-c4e1-11e5-9639-7bb11e5f0193.png)

It overflows into the second screen:
![untitled _011](https://cloud.githubusercontent.com/assets/2535481/12609617/2821e82a-c4e3-11e5-9861-435e097ae3f1.png)

It looks normal when only few windows are open (same system, other workspace):
![workspace 2_012](https://cloud.githubusercontent.com/assets/2535481/12609680/986f2bc4-c4e3-11e5-8811-330dc31e40bb.png)

This also happens in single-monitor setups. Some themes (Metro, Droit) use a bit more space so fewer icons can be displayed before things start overlapping but it happens with the default Cinnamon theme too.

Idea 1: Show a pager in the window list when there is not enough space for all window icons. And even when there's just enough space, I'd rather switch back and forth and see window titles than have all icons there at once without any space left for the title texts.

Idea 2: Allow the panel to have 2 rows.

**Comment by keitalbame** [comment]
Although I think idea 2 would be a feature that I would use, in my case I'm not seeing that behaviour with 15 applications open (with icon and titles).

Could be that the applet with the numbers (virtual desktops!?) could be the reason? Did you tried to remove and test?

I tested on a laptop with single monitor, with cinnamon and other themes.

What is your setup?

**Comment by basic6** [comment]
The applet with the numbers is the standard "Workspace switcher" applet. In the screenshot, you can also see the "System Monitor" applet.
But it doesn't matter if you're using these specific applets or not. It's simply a matter of how much space is left for the window list. If you don't use those additional applets, then you simply have to open more windows before the panel starts growing.

Try this:
Open leafpad or gedit (or really any program that allows you to open new windows). Use Ctrl + N to open new windows - not just one, hold Ctrl + N to keep opening more and more new windows. At some point, the window list will be full and the panel will grow (instead of showing a pager). Watch the bottom right corner of your screen, the Calendar applet (if you use that as a clock) is probably the first to disappear.

The whole point is that the panel and/or its window list does not seem to have any paging mechanism, so as soon as there's not enough space to show all window buttons, the whole panel grows, parts of it move outside the viewport and some applets even overlap.
Since there really isn't that much space available for the window list in my setup (because of the other two applets), this happens a *lot*. And because it is so frustrating when applets overlap so that they can't be used anymore and tray icons disappear, it makes me want to switch to another desktop environment.

**Comment by keitalbame** [comment]
#4964 also reporting same issue.

**Comment by Ocelot92** [comment]
I can confirm this on Linux Mint 17.3. I have the default taskbar, no desktop switcher or else.

**Comment by JosephMcc** [comment]
It's simply an issue that when there are too many windows open in the window list we don't do anything to handle the overflow. Simple issue, simple to reproduce but likely not so simple to fix.

**Comment by Ocelot92** [comment]
Yep, i know.

**Comment by TooHuman** [comment]
This is still an issue in Linux Mint 18 and 18.1.

**Comment by cmrdt** [comment]
although not a complete fix arent themes a factor?
crunchy-green is slimmer, therefore a bit more breathing-space when space is cramped
and more headroom-space-distance to full 


to deal with this overlap-overflow
if it happens maybe switch z-order when hover and press shift|alt ?? 
- (similar to how inkscape handles overlapping objects)

**Comment by Cpt-Derp** [comment]
Any news on this? Tripped on this bug with wide padding and small resolutions.

I tried looking at the JS of the applet to see if there's any way the applet can detect its end coordinate, and detect the start coordinate of the adjacent applet, to perhaps draw scroll buttons on either side when the two overlap, but I haven't had luck.

**Comment by crobarcro** [comment]
Just came here to say I see the same issue on my laptop with Mint 19. The workspace switching icons overlap the window icons, making them impossible to click (have to use Alt+tab to access).

**Comment by ShamrockLee** [comment]
It happens on my Debian bullseye (testing) when the "System Settings > Themes > Themes > Desktop" option is set to "Minty", "New-Minty" and "BlackMATE", but not "Blue-Submarine", "Green-Submarine", "Menta", "BlueMenta" or the default "cinnamon".
![Themes setting window and the Desktop options](https://user-images.githubusercontent.com/44064051/68547779-795faf80-0420-11ea-9b1e-d4c8a1f17a1a.png)
I filed [issue 597 of cinnamon-spices-themes](https://github.com/linuxmint/cinnamon-spices-themes/issues/597), and germanfr, the author of New-Minty replied that he/she cannot reproduce it and that it could be a Cinnamon bug.
Issues that might be related: #4458, #686

**Comment by Cpt-Derp** [comment]
This is a consistently reproducible bug happening with multiple themes. 

It would seem some applets are pushovers but others aren't. Either way, the border between the center and right regions of the panel is being pushed too far to the right when the window list expands, and instead of drawing scroll bars, it pushes applets off screen and draws over other applets that stood in its way.

**Comment by Cpt-Derp** [comment]
Reproduced with Minty and New-Minty
![Screenshot from 2019-11-10 13-14-17](https://user-images.githubusercontent.com/3199836/68548593-8ba80200-03bc-11ea-836a-3e68c519aa53.png)
![Screenshot from 2019-11-10 13-12-55](https://user-images.githubusercontent.com/3199836/68548594-8ba80200-03bc-11ea-88ec-7d98ec98843c.png)

**Comment by Cpt-Derp** [comment]
@ShamrockLee Default theme too

![Screenshot from 2019-11-10 13-20-22](https://user-images.githubusercontent.com/3199836/68548615-e5a8c780-03bc-11ea-9c46-3482d2ca1f84.png)

**Comment by Cpt-Derp** [comment]
@germanfr Reproduced with regular window list and it's quite a sight to behold.
![Screenshot from 2019-11-10 13-46-31](https://user-images.githubusercontent.com/3199836/68548971-ee9b9800-03c0-11ea-9ae2-8d5547419545.png)

**Comment by germanfr** [comment]
I **didn't** reproduce it with the regular window list applet @Cpt-Derp. GWL only and with all themes. It's not related to New-Minty.

**Comment by Cpt-Derp** [comment]
> I **didn't** reproduce it with the regular window list applet @Cpt-Derp. GWL only and with all themes. It's not related to New-Minty.

Guess I wasn't clear. I wasn't showing it has anything to do with New-Minty, just that the same bug affects the regular window list, and I could at least reproduce it. My apologies.

**Comment by ShamrockLee** [comment]
> @ShamrockLee Default theme too
> 
> ![Screenshot from 2019-11-10 13-20-22](https://user-images.githubusercontent.com/3199836/68548615-e5a8c780-03bc-11ea-9c46-3482d2ca1f84.png)

You're right. They all have this bug. The different thing is the tolerable number of windows. That is, 8 windows are enough to push the applets right in Minty and New-Minty, but I need 23 windows to make it happen to Menta, 27 to the cinnamon.

**Comment by mtwebster** [comment]
~Fixed~ Mitigated in 566cad571e2bf6c9bcedf and 27226e616586af251

A complete fix would require more changes than I'm prepared to make for now (the chance of regression would be severe and I don't want that in 5.2). Probably some sort of overflow button at the end of the list to collect items that don't fit, or making it scrollable somehow.

On the other hand, though the issue can still be reproduced, it takes a lot more windows to do so (around 50 or so depending on how crowded the panel already is).

**Comment by Cloudwalk9** [comment]
@mtwebster

Praise be. If this issue is properly fixed (as well as Muffin's performance issues being fixed), I'd switch back to Cinnamon on every distro I touch.

Thank you so much for at least mitigating it. This was a serious issue on smaller resolutions in my experience.

**Comment by sharif-jobayed** [comment]
Any update on this issue yet..?

**Comment by metafarion** [comment]
I'm suffering from this as well.  If it's really hard to implement a full-on overflow section for the window list, what about just making the overflow disappear behind the tray area rather than pushing it off the side or visibly overlapping.  We'd lose the ability to switch to certain windows with a click, but we can still use Alt+Tab, whereas the system tray loses most of its functionality if you can't see it.

Basically, I'd rather be able to see my clock than my 34th open window in the panel.

**Comment by mtwebster** [comment]
@metafarion what cinnamon  version? This should no longer be an issue in 5.4, for at least the 'stock' window lists.

**Comment by metafarion** [comment]
Oh you're right, I'm on 5.2; I didn't notice that 5.4 is still masked on Gentoo.  I guess I have a few hoops to jump through.

**Comment by metafarion** [comment]
Yeppppp, ok works great in 5.4, thanks!
